tests/data_classes.py

```diff
@@ -0,0 +1,647 @@
+        self.zfs = ZFS() if not zfs else zfs
+        assert isinstance(self.zfs, ZFS) is True
+
+    def __repr__(self):
+        return self.name
+
+    def convert_to_row(self, **kwargs):
+        raise NotImplemented
+
+
+class Snapshot(Resource):
+
+    def __init__(self, name, parent_dataset, zfs=None):
+        super().__init__(name, zfs)
+        self.parent = parent_dataset
+        if isinstance(self.parent, str):
+            self.parent = Jail(self.parent)
+        if self.exists:
+            for k, v in self.zfs.get_snapshot_safely(self.name).items():
+                setattr(self, k, v)
+
+    @property
+    def exists(self):
+        return self.zfs.get_snapshot_safely(self.name) is not None
+
+    def convert_to_row(self, **kwargs):
+        full = kwargs.get('full', False)
+        if full:
+            name = self.id
+        else:
+            name = self.id.split('@')[-1]
+            # If snap is of root dataset - we add /root
+            if self.id.split('/')[-1].split('@')[0] == 'root':
+                name += '/root'
+
+        return Row({
+            'name': name,
+            'created': self.properties['creation']['value'],
+            'rsize': self.properties['referenced']['value'],
+            'used': self.properties['used']['value']
+        })
+
+
+class Release(Resource):
+
+    @property
+    def exists(self):
+        return self.name in [
+            r['name'].split('/')[-1]
+            for r in self.zfs.releases_dataset['children']
+        ]
+
+    def convert_to_row(self, **kwargs):
+        return Row({'release': self.name})
+
+
+# A simple class which can tell us the state of the jail
+class Jail(Resource):
+
+    def convert_to_row(self, **kwargs):
+        short_name = kwargs.get('short_name', True)
+        full = kwargs.get('full', False)
+
+        props = self.jail_dataset['properties']
+
+        ip4 = self.config.get('ip4_addr', 'none')
+        if self.config.get('dhcp', 'off') == 'on':
+            ip4 = 'DHCP'
+
+        if ip4 == 'none':
+            ip4 = '-'
+
+        if full:
+            release = self.release
+        else:
+            release = '-'.join(self.release.rsplit('-')[:2])
+
+        template = '-'
+        jail_origin = self.root_dataset['properties']['origin']['value']
+        if jail_origin:
+            template = jail_origin.rsplit(
+                '/root@', 1
+            )[0].rsplit('/', 1)[-1]
+            if any(
+                    v in template.lower()
+                    for v in ('release', 'stable')
+            ):
+                template = '-'
+
+        return Row({
+            'name': self.short_name if short_name else self.name,
+            'jid': self.jid,
+            'state': 'up' if self.running else 'down',
+            'boot': self.config.get('boot', 'off'),
+            'type': 'jail' if not self.is_template else 'template',
+            # TODO: Add support for plugins
+            'ip6': '-',  # FIXME: Change when ip6 tests are added
```

Comment:
Use `'none'` to be like the other props that aren't specified
```suggestion
             'ip6': 'none',  # FIXME: Change when ip6 tests are added
```

Reply:
Umm this is to ensure matching output with the listing of jails etc. We don't use this for any other reason. If we change it to none, then the comparison will fail

Reply:
Got it :+1: